### DISCUSSION TITLE
Clangd parsing occasionally failed.

### DISCUSSION THREAD

**Discussion by zx616456607** [clangd/clangd#1958] [Q&A]
Clangd serverOptions sets the installation path for clangd.exe, and options.cwd sets the workspace directory.

Clangd clientOptions sets the configuration fallbackFlags of the header file. These configuration items I configure are the parent directory of the header file directory, and the header file is introduced through the relative path in the editor.

When opening and switching files, we will dynamically set the header file configuration and reset the header file configuration through the clangd restart command.

We are currently doing this and have encountered some problems, such as incomplete display of reference files when viewing references, blank page of call tree information, etc.

![clangd restart](https://github.com/clangd/clangd/assets/19658190/186046d5-1339-4ca1-a452-de7c73f11f66)
![clangd fallbackFlags](https://github.com/clangd/clangd/assets/19658190/cd0b0f7e-8b6c-4410-b518-357c2161a5c8)

**Comment by HighCommander4** [comment]
> We are currently doing this and have encountered some problems, such as incomplete display of reference files when viewing references, blank page of call tree information, etc.

Can you provide a code example and [clangd logs](https://clangd.llvm.org/troubleshooting.html#gathering-logs)?

Also, what is the relevance of the screenshots of the vscode-clangd source code that you are showing?

**Reply by zx616456607** [reply]
Open source/a.cpp first.
The contents of the document are as follows:
`#include "include/a.h"
int fn(int a) { 
  int tmp = a;
  return tmp;
}`
![打开a cpp](https://github.com/clangd/clangd/assets/19658190/3ed51cae-08e8-4109-b90e-8618a01b51e3)


Open include/a.h again.
The contents of the document are as follows:
`int fn(int a);`
There is an error in the log: Failed to find compilation database.....

![打开a h](https://github.com/clangd/clangd/assets/19658190/1efeedbd-6a07-4e25-8869-e93405b497f1)


View the call hierarchy
![查看调用树](https://github.com/clangd/clangd/assets/19658190/2448e0ed-6080-4f59-bd8d-33598e52be9b)


Header file configuration, passed to clangd server through vscode-clangd fallbackFlags

In the header file configuration, the first configuration is the superior directory of A.H and a.cpp

![头文件配置](https://github.com/clangd/clangd/assets/19658190/57b6e96c-1cf8-4129-989e-f03c4fb84589)

**Reply by HighCommander4** [reply]
I think your issue may be that call hierarchy relies on the project's index to work, and project needs to have a compilation database (`compile_commands.json` file) in order for clangd to create the index.

Have a look at https://clangd.llvm.org/installation#compile_commandsjson for some ways of generating a compilation database.

**Reply by zx616456607** [reply]
Excuse me, if this file is available, do I have to pass parameters to clangd server through clangd client? Or what should I do?

**Reply by HighCommander4** [reply]
Not sure I understand.

Clangd looks for `compile_commands.json` in the directory of the opened file and its ancestor directories, or you can specify a custom directory using the `--compile-commands-dir` command-line parameter to clangd.

**Reply by zx616456607** [reply]
Thank you. I'll try.

**Reply by Logger1106** [reply]
Using the VS clangd extension of the call hierarchy, the call hierarchy of function a was first queried, and then when checking function b, the log still shows the call hierarchy of the requested function a, which displays the call hierarchy of function a.
